linkerd/app/inbound/src/direct.rs

```diff
@@ -25,14 +25,23 @@ pub struct RefusedNoIdentity(());
 #[error("a named target must be provided on gateway connections")]
 struct RefusedNoTarget;
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct LocalTcp {
+    addr: Remote<ServerAddr>,
+    client_id: tls::ClientId,
+    permit: policy::Permit,
+}
+
 #[derive(Debug, Clone)]
-pub(crate) struct Local {
+pub(crate) struct LocalHttp {
     addr: Remote<ServerAddr>,
     policy: policy::AllowPolicy,
     client: ClientInfo,
-    protocol: Option<SessionProtocol>,
+    protocol: SessionProtocol,
 }
 
+type Local = svc::Either<LocalTcp, LocalHttp>;
```

Comment:
This is neat! So this basically allows us to just return the inner predicate on L129 and have the switch layer in one line? That's really cool.